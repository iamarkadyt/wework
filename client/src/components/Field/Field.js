import React from 'react'
import DropdownList from 'react-widgets/lib/DropdownList'
import DateTimePicker from 'react-widgets/lib/DateTimePicker'
import Multiselect from 'react-widgets/lib/Multiselect'
import Moment from 'moment'
import momentLocalizer from 'react-widgets-moment'
import './Field.scss'
import { node, string, func, object, bool, number, arrayOf, oneOfType, instanceOf } from 'prop-types'

Moment.locale('en')
momentLocalizer()

const field = ({
    // different fields require different props
    type,
    name, value, onChange, // all except submit and button
    placeholder, error, // all except checkbox, submit and button
    label, // checkbox, submit, button
    list, // list, multiselect
    rows, // textarea
    disabled, // date
    onClick, style, // button, linkButton, textarea
    inline, // container
    containerStyle, // container
    children
}) => {
    let field = null
    const commonProps = { name, id: name, value, onChange }

    switch (type) {
        case "textarea":
            field = <textarea
                {...commonProps}
                placeholder={placeholder || 'Say something...'}
                className="rw-widget-container rw-widget-input"
                style={{
                    padding: '.5rem',
                    fontSize: '1rem',
                    width: '100%',
                    resize: 'none',
                    ...style
                }}
                rows={rows} />
            break
        case "date":
            field = <DateTimePicker
                {...commonProps}
                placeholder={placeholder || 'Pick a date...'}
                time={false}
                format="MMM DD, YYYY"
                editFormat="M/DD/YYYY"
                disabled={disabled} />
            break
        case "list":
            field = <DropdownList
                {...commonProps}
                placeholder={placeholder || "Choose one..."}
                data={list} />
            break
        case "multiselect":
            field = <Multiselect
                {...commonProps}
                placeholder={placeholder || "Enter a few items..."}
                data={list} />
            break
        case "checkbox":
            field = <label htmlFor={name}>
                <input type="checkbox" {...commonProps} />
                &nbsp;&nbsp;{label}
            </label>
            break
        case "submit":
            field = <button
                type="submit"
                className="button-action">
                {label || children}
            </button>
            break
        case "button":
            field = <button
                className="button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        case "linkButton":
            field = <button
                className="link-button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        default:
            field = <input
                {...commonProps}
                placeholder={placeholder}
                className="rw-widget-container rw-widget-input"
                type={type} />
    }

    return (
        <div style={containerStyle}
            className={['Field-container', inline ? 'inline' : ''].join(' ')}>
            {label
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <label htmlFor={name}>{label}</label>}
            {field}
            {error
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <span className="error">{error}</span>}
        </div>
    )
}

field.propTypes = {
  type: string,
  name: string,
  value: oneOfType([string, arrayOf(string), bool, instanceOf(Date)]),
  onChange: func,
  placeholder: string,
  error: string,
  label: string,
  list: arrayOf(string),
  rows: oneOfType([string, number]),
  disabled: bool,
  onClick: func,
  style: object,
  inline: bool,
  containerStyle: object,
  children: oneOfType([arrayOf(node), node])
}

export default field
