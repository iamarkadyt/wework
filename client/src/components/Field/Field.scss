@import 'variables.scss';
.Field-container {
    &.inline {
        display: inline;
    }
    width: 100%;
    margin: 1rem 0 0 0;
    input[type=checkbox] {
        transform: scale(1.5);
    }
    .button,
    button[type=submit] {
        min-width: 6rem;
        border-radius: .2rem;
        font-size: 1.1rem;
        padding: .2rem 1rem;
        border-color: $color-button-border;
        background: $color-button-backround;
        color: $color-button-font;
        &:active {
            border-color: $color-button-active-border;
            background: $color-button-active-background;
            color: $color-button-active-font;
        }
        &:hover:not(:active) {
            border-color: $color-button-hover-border;
            background: $color-button-hover-backround;
            color: $color-button-hover-font;
        }
    }
    button[type=submit] {
        float: right;
    }
    .button-action {
        background-color: darken(cornflowerblue, 10%);
        border-color: darken(cornflowerblue, 10%);
        background: linear-gradient(to bottom, darken(cornflowerblue, 10%), darken(cornflowerblue, 20%));
        color: white;
        &:active {
            background: linear-gradient(to bottom, darken(cornflowerblue, 25%), darken(cornflowerblue, 15%));
        }
        &:hover:not(:active) {
            background: linear-gradient(to bottom, darken(cornflowerblue, 8%), darken(cornflowerblue, 18%));
        }
    }
    .link-button {
        background: none;
        border: none;
        font-weight: bold;
        font-size: 1.1rem;
        color: $color-font-secondary;
        &:hover {
            color: $color-font-main;
            cursor: pointer;
        }
    }
    input[type=text],
    input[type=password] {
        font-size: 1rem;
        padding: .6rem;
        width: 100%;
    }
    label {
        display: inline-block;
        padding: .2rem;
    }
    textarea {
        background-color: $color-field-background;
        border-color: $color-field-border;
        color: $color-field-font;
    }
    .error {
        display: inline-block;
        color: $color-font-error;
        padding: .2rem;
    }
}