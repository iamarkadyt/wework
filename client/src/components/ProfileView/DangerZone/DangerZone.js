import React from 'react'
import Field from '../../Field/Field'
import { func, bool } from 'prop-types'

const dangerZone = ({
  navTo,
  quitEntryDeletingMode,
  profileBelongsToAuthedUser
}) => {
  return profileBelongsToAuthedUser ? (
    <section>
        <div />
        <Field
            type="button"
            label="Delete Profile"
            inline
            style={{
                backgroundColor: 'red',
                borderColor: 'red',
                color: 'white'
            }}
            containerStyle={{ margin: 0 }}
            onClick={() => {
                quitEntryDeletingMode()
                navTo('delete-profile')
            }} />
    </section>
  ) : null
}

dangerZone.propTypes = {
  navTo: func.isRequired,
  profileBelongsToAuthedUser: bool.isRequired,
  quitEntryDeletingMode: func.isRequired
}

export default dangerZone
