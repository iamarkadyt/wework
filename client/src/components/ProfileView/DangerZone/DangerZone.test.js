import React from 'react'
import DangerZone from './DangerZone'
import Field from '../../Field/Field'
import { shallow } from 'enzyme'
import cloneDeep from 'lodash.clonedeep'

describe('DangerZone', () => {
  let mountedComponent, props

  const getMockProps = () => {
    return cloneDeep({
      navTo: jest.fn(),
      quitEntryDeletingMode: jest.fn(),
      profileBelongsToAuthedUser: true
    })
  }

  const comp = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(<DangerZone {...props} />)
    } 
    return mountedComponent
  }

  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()
  })

  describe('props', () => {
    describe('effects', () => {
      describe('profileBelongsToAuthedUser', () => {
        it('if false, comp renders nothing', () => {
          props.profileBelongsToAuthedUser = false
          expect(comp().find('section').exists()).toBe(false)
        })

        it('if true, comp renders everything', () => {
          props.profileBelongsToAuthedUser = true
          expect(comp().find('section').exists()).toBe(true)
        })
      })
    })
  })

  describe('functions', () => {
    it('calls quitEntryDeletingMode', () => {
      comp().find(Field).simulate('click')
      expect(props.quitEntryDeletingMode).toHaveBeenCalled()
    })

    it('calls navTo', () => {
      comp().find(Field).simulate('click')
      expect(props.navTo).toHaveBeenCalled()
    })

    it('calls navTo with correct args', () => {
      comp().find(Field).simulate('click')
      expect(props.navTo).toHaveBeenCalledWith('delete-profile')
    })
  })

  it('matches snapshot', () => {
    expect(comp()).toMatchSnapshot()
  })
})
