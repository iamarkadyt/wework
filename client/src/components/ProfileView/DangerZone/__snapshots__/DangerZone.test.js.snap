// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DangerZone matches snapshot 1`] = `
<section>
  <div />
  <field
    containerStyle={
      Object {
        "margin": 0,
      }
    }
    inline={true}
    label="Delete Profile"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "red",
        "borderColor": "red",
        "color": "white",
      }
    }
    type="button"
  />
</section>
`;
